### PR TITLE
new InputAddressSelect component

### LINKED ISSUE #3008: Address inputs in non edit mode should accept addresses
Warning: The following issue can lead to disasters for normal users.
Disaster = unexpected replacement of the address they typed by 0x0 and of contracts with 0x0 instead of the address they intended.

The address input has 2 modes:
- edit (we can type 0x124...)
- search (we can type Bob and search for the account called Bob...)

In various places, I noticed that the mode is set by default to either edit or search depending on the context. That´s fine.

I would propose the following:
- if the address input component is in 'search' mode
- and if the user type in a valid address that is not matching a known account
- then switch for the user to the 'edit' mode

Not doing so is confusing for the user who types a valid address.
The user would see the following:
screen shot 2016-10-31 at 13 04 17
which looks OK for normal users.

However, since we are in 'search' mode and the address is a good one but not in the list of know addresses, the address that will be used when calling the contract **will not** be the one the user sees, but 0x0 instead.

### PR DESCRIPTION
**Pull request by derhuerst**
This PR gives `InputAddressSelect` an overhaul. It fixes #3008. #3006 gets fixed indirectly because by default the input is empty.

The new idea behind the rewrite was to make it as comfortable as possible to fill in an address. There's completion by name & address and no toggle. You can enter an address both *with* and *without* a leading address.

What #3057 does for `InputAddress`, is done by this PR for `InputAddressSelect`.

Edit: I haven't checked for compatibility in all the places where `InputAddressSelect` is used. Will do that tomorrow.

### CODE REVIEW

**Comment by derhuerst** [comment]
Let me know what you think about the default `IdentityIcon` being gray. (It will show the pattern once the input begins with `0x`.)

**Comment by jacogr** [comment]
Marked in-progress until you are comfortable and it actually has been checked. The risk is high, do a proper testing job on this one first before PR-ing.

**Comment by jacogr** [comment]
As per @ngotchac PR - unless the address is valid, no IdentityIcon at all. It only gets displayed when the address is valid.

**Comment by chevdor** [comment]
Tested in the current version. I like the direction this is taking.

I miss the ability to click on the control to pull the list of known addresses and select from there.
While no absolute must, the previous resolution from a given address to a known name was nice and is (so far) missing in what I tested.

The toggle edit/search disappeared, which IMHO is a GREAT thing! Well done @derhuerst.
Warning, in the version I tested, the validation was gone: 0x12234 was not seen as invalid address, 0xasdf either.

I am currently working on an address component (not react), here are some screenshots if that can help:
screen shot 2016-11-02 at 00 31 21
screen shot 2016-11-02 at 00 31 33
screen shot 2016-11-02 at 00 30 55
screen shot 2016-11-02 at 00 30 27

**Comment by chevdor** [comment]
@jacogr in the current version, this is not the case:
screen shot 2016-11-02 at 00 34 53

**Comment by derhuerst** [comment]
> Warning, in the version I tested, the validation was gone: 0x12234 was not seen as invalid address, 0xasdf either.

Will address.

**Comment by gavofyork** [comment]
> address both with and without a leading address

???

**Comment by derhuerst** [comment]
>> address both with and without a leading address
>???

* You can enter an address both with and without a leading `0x`.

**Comment by chevdor** [comment]
See #3093

**Comment by jacogr** [comment]
Comments -

- UI is now inconsistent, this field doesn't not have the placeholders or labels in the correct places
- looking at 2 icon fields side-by-side - they don't match up

inconsistent-inputs

- Own accounts always first, then addresses, then contracts

own-first

- Invalid addresses still show IdentityIcons

invalid-error-identityicon

- I'm assuming your comment from yesterday is not valid anymore? (re the testing)

**Comment by ngotchac** [comment]
I like better the icon outside of the input as @derhuerst did. I think it should be like that for every input with an icon.

**Comment by jacogr** [comment]
I don't have a preference either way - but if it gets moved, it gets moved everywhere - needs to be consistent, cannot do it half-way.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by jacogr** [comment]
Address input still don't look the same in the UI. Compare the first to the second. It is no doubt better than the initial versions (i.e. gets closer), but still misses the spot.

parity 2016-11-02 18-37-23

**Comment by derhuerst** [comment]
> Address input still don't look the same in the UI. Compare the first to the second. It is no doubt better than the initial versions (i.e. gets closer), but still misses the spot.

The `IdentityIcon` for an invalid address is not `32px` wide yes. That's the case everywhere and should be addresses separately.

If you mean the underline of the input, I strongly oppose making that look like the one above (which looks broken). Only doing it for the sake of pixel-perfect consistency is ridiculous, especially because we should rewrite/refactor `AddressSelect` soon.

If you mean something else, I'm happy to hear it. Generally speaking, those controls don't *feel* the same anyways, as the look completely different once you clicked them (which is on purpose).

**Comment by derhuerst** [comment]
To stop this bikeshedding… I'll have a look at how much work it is to replace `AddressSelect` by `InputAddressSelect` (or rather merge them).

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by jacogr** [comment]
Updated. @derhuerst was 100% correct in his assessment - instead of re-inventing the wheel, rather use the AddressSelect component as developed by @ngotchac and adapt it to also take care of "copy text" inputs. Basically the code as originally developed is 90% the same, just re-done - here where have no case of repeating the same code again. No styling issues, already taken care of in the original.

In short: Simplified greatly. Video of it in action -

https://youtu.be/vudv7NFw0-o

Excuse the - (a) slugishness, Parity is syncing and (b) crash of Parity (stack-trace in @retreat)

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by jacogr** [comment]
Updated. Closes https://github.com/ethcore/parity/issues/3073 - contracts in the address search.

https://youtu.be/_ooiw0FysVk

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ec2d3ecc27eaf4aeed9be366847d785fd1fe8ee8
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by chevdor** [comment]
Should be used for #3134
